### ISSUE TITLE
Spectral OAS2 validation throws a non-existent error without path value

### ISSUE THREAD

**Issue by jerzyn** [stoplightio/spectral#109]
Running spectral OAS2 rules and functions sets on a OAS2 `.yaml` file on all tests (2 different `yaml files`) results in a cryptic error without a `"path"` parameter:
```
{
    "name": "oas2-schema",
    "summary": "Validate structure of OpenAPIv2 specification.",
    "message": "should NOT have additional properties: 0",
    "path": [],
    "severity": 50,
    "severityLabel": "error"
}
```
Those files don't throw any errors in any other validation tool (swagger editor, apiary, etc.). I am happy to provide them if needed.

### **I'm submitting a...**
  - [X] bug report
  - [ ] feature request

### What is the current behavior?

Running:
```
spectral.addFunctions(oas2Functions());
spectral.addRules(oas2Rules());
const results = spectral.run(myOAS);
```
on a OAS2 `.yaml` file

### What is the expected behavior?
Do not throw this error or be more specific on it's origin and place in the file.

### What is the motivation / use case for changing the behavior?
User experience and compliance with OAS2 specification

### Please tell us about your environment:

  - Version: @stoplight/spectral@1.2.0
  - Framework: [ ]
  - Language: [all]

### Other information

It also throws 
```
oasOpParams expects a resolved object, but none was provided. Results may not be correct.
oasPathParam expects a resolved object, but none was provided. Results may not be correct.
```
which does not say anything of value without more explanation.

**Comment by philsturgeon** [comment]
@jerzyn hey there, let's get this sorted out. 

First up, don't worry about the "expects a resolved object, but none was provided. Results may not be correct." warnings, they'll go away in v2.0 which is pretty close to being ready (check the release/2.0 branch if you are interested).

Could I have your original YAML please? You can gist the whole thing, obfuscate some bits, or trim it down to the smallest possible OAS file that recreates the issue. From there I can try to replicate the issue.

**Comment by jerzyn** [comment]
thanks! Do you need both files or one is enough? It will take me a minute to 'anonymize' it.

**Comment by jerzyn** [comment]
here is one:
https://gist.github.com/jerzyn/d0201558bf52b3659c0d43cac646896c
Let me know if you need the other one, however it is a beast of a spec...

**Comment by philsturgeon** [comment]
It was an or list, not an and list! 😁

I’ll take a look at this example in the morning and see if I can recreate anything.

**Comment by philsturgeon** [comment]
Hey! So, good news and bad news @jerzyn. I have got the release/2.0 branch CLI taking a look at your spec and it has a _lot_ of opinions.

```bash
$ git checkout release/2.0
$ yarn setup && yarn build
$ ./dist/bin/run lint https://gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml

OpenAPI 2.0 (Swagger) detected
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews-stats,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'valid-example' on node at path '$,definitions,halRoot':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProducts':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProduct':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productPatch':
Error: can't resolve reference #/definitions/product from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviews':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReview':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewsStats':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewStatsItem':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halUser':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,userNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishlist':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishProduct':
Error: can't resolve reference #/definitions/halResource from id #

/Users/phil/src/spectral/https:/gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml
   1:1   warning  api-host               OpenAPI `host` must be present and non-empty string
   2:6   warning  info-contact           Info object should contain `contact` object
  55:16  warning  operation-tags         Operation should have non-empty `tags` array
  55:16  warning  operation-operationId  Operation should have an `operationId`
  55:16  warning  operation-description  Operation `description` must be present and non-empty string
  57:9   warning  operation-operationId  Operation should have an `operationId`
  57:9   warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-operationId  Operation should have an `operationId`
  69:16  warning  operation-description  Operation `description` must be present and non-empty string
  71:9   warning  operation-operationId  Operation should have an `operationId`
  71:9   warning  operation-tags         Operation should have non-empty `tags` array
 124:10  warning  operation-operationId  Operation should have an `operationId`
 124:10  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-operationId  Operation should have an `operationId`
 146:16  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-description  Operation `description` must be present and non-empty string
  156:9  warning  operation-operationId  Operation should have an `operationId`
  156:9  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-description  Operation `description` must be present and non-empty string
 199:16  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-description  Operation `description` must be present and non-empty string
 259:16  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-description  Operation `description` must be present and non-empty string
 305:16  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-tags         Operation should have non-empty `tags` array
 329:16  warning  operation-operationId  Operation should have an `operationId`
 329:16  warning  operation-description  Operation `description` must be present and non-empty string
 329:16  warning  operation-tags         Operation should have non-empty `tags` array
  339:9  warning  operation-operationId  Operation should have an `operationId`
  339:9  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-operationId  Operation should have an `operationId`
 352:16  warning  operation-description  Operation `description` must be present and non-empty string
  362:9  warning  operation-tags         Operation should have non-empty `tags` array
  362:9  warning  operation-operationId  Operation should have an `operationId`
 374:12  warning  operation-tags         Operation should have non-empty `tags` array
 374:12  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-tags         Operation should have non-empty `tags` array
 385:16  warning  operation-description  Operation `description` must be present and non-empty string
 402:10  warning  operation-operationId  Operation should have an `operationId`
 402:10  warning  operation-tags         Operation should have non-empty `tags` array
 423:12  warning  operation-operationId  Operation should have an `operationId`
 423:12  warning  operation-tags         Operation should have non-empty `tags` array
  744:8  warning  valid-example          Examples must be valid against their defined schema

✖ 58 problems (0 errors, 58 warnings, 0 infos)
```

So the TypeError and Error messages at the top looks like bugs, but it also seems that Spectral will keep on trucking past those bugs in order to provide linting regardless.

If you would like to work with the CLI version for now following the instructions above, you can use those to lint, whilst we get these bugs sorted out over in #111.

**Comment by philsturgeon** [comment]
Hey! So, good news and bad news @jerzyn. I have got the release/2.0 branch CLI taking a look at your spec and it has a _lot_ of opinions.

```bash
$ git checkout release/2.0
$ yarn setup && yarn build
$ ./dist/bin/run lint https://gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml

OpenAPI 2.0 (Swagger) detected
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/products/{product_id}/reviews-stats,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist,x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'operation-2xx-response' on node at path '$,paths,/users/{user_id}/wishlist/{product_id},x-summary':
TypeError: Cannot convert undefined or null to object
Encountered error when running rule 'valid-example' on node at path '$,definitions,halRoot':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProducts':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halProduct':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productPatch':
Error: can't resolve reference #/definitions/product from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,productNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviews':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReview':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewsStats':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halReviewStatsItem':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halUser':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,userNotFound':
Error: can't resolve reference #/definitions/problemDetail from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishlist':
Error: can't resolve reference #/definitions/halResource from id #
Encountered error when running rule 'valid-example' on node at path '$,definitions,halWishProduct':
Error: can't resolve reference #/definitions/halResource from id #

/Users/phil/src/spectral/https:/gist.githubusercontent.com/jerzyn/d0201558bf52b3659c0d43cac646896c/raw/0bc6e80d7bce9d8556f791a641924beb624e2f7b/swagger-demo.yaml
   1:1   warning  api-host               OpenAPI `host` must be present and non-empty string
   2:6   warning  info-contact           Info object should contain `contact` object
  55:16  warning  operation-tags         Operation should have non-empty `tags` array
  55:16  warning  operation-operationId  Operation should have an `operationId`
  55:16  warning  operation-description  Operation `description` must be present and non-empty string
  57:9   warning  operation-operationId  Operation should have an `operationId`
  57:9   warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-tags         Operation should have non-empty `tags` array
  69:16  warning  operation-operationId  Operation should have an `operationId`
  69:16  warning  operation-description  Operation `description` must be present and non-empty string
  71:9   warning  operation-operationId  Operation should have an `operationId`
  71:9   warning  operation-tags         Operation should have non-empty `tags` array
 124:10  warning  operation-operationId  Operation should have an `operationId`
 124:10  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-operationId  Operation should have an `operationId`
 146:16  warning  operation-tags         Operation should have non-empty `tags` array
 146:16  warning  operation-description  Operation `description` must be present and non-empty string
  156:9  warning  operation-operationId  Operation should have an `operationId`
  156:9  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-tags         Operation should have non-empty `tags` array
 176:11  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-operationId  Operation should have an `operationId`
 199:16  warning  operation-description  Operation `description` must be present and non-empty string
 199:16  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-tags         Operation should have non-empty `tags` array
  209:9  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-operationId  Operation should have an `operationId`
 236:10  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-tags         Operation should have non-empty `tags` array
 259:16  warning  operation-description  Operation `description` must be present and non-empty string
 259:16  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-operationId  Operation should have an `operationId`
  269:9  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-description  Operation `description` must be present and non-empty string
 305:16  warning  operation-tags         Operation should have non-empty `tags` array
 305:16  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-operationId  Operation should have an `operationId`
 307:10  warning  operation-tags         Operation should have non-empty `tags` array
 329:16  warning  operation-operationId  Operation should have an `operationId`
 329:16  warning  operation-description  Operation `description` must be present and non-empty string
 329:16  warning  operation-tags         Operation should have non-empty `tags` array
  339:9  warning  operation-operationId  Operation should have an `operationId`
  339:9  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-tags         Operation should have non-empty `tags` array
 352:16  warning  operation-operationId  Operation should have an `operationId`
 352:16  warning  operation-description  Operation `description` must be present and non-empty string
  362:9  warning  operation-tags         Operation should have non-empty `tags` array
  362:9  warning  operation-operationId  Operation should have an `operationId`
 374:12  warning  operation-tags         Operation should have non-empty `tags` array
 374:12  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-operationId  Operation should have an `operationId`
 385:16  warning  operation-tags         Operation should have non-empty `tags` array
 385:16  warning  operation-description  Operation `description` must be present and non-empty string
 402:10  warning  operation-operationId  Operation should have an `operationId`
 402:10  warning  operation-tags         Operation should have non-empty `tags` array
 423:12  warning  operation-operationId  Operation should have an `operationId`
 423:12  warning  operation-tags         Operation should have non-empty `tags` array
  744:8  warning  valid-example          Examples must be valid against their defined schema

✖ 58 problems (0 errors, 58 warnings, 0 infos)
```

So the TypeError and Error messages at the top looks like bugs, but it also seems that Spectral will keep on trucking past those bugs in order to provide linting regardless.

If you would like to work with the CLI version for now following the instructions above, you can use those to lint, whilst we get these bugs sorted out over in #111.

**Comment by jerzyn** [comment]
thanks @philsturgeon Looks like most of those `errors` are just warnings about including stuff like operationId or the inability of spectral to resolve some references. The first ones are about interpreting `x-summary` vendor extension as a part of the spec, which it should just ignore IMHO. Interestingly it does not complain about the same stuff as spectral itself (the one I have included in in the issue). Do you know what may be the cause of it? I am referring to the `should NOT have additional properties: 0` error.
About the CLI - how can I use it? Any docs (I assume i have to switch to the `cli` branch first) on the usage?

**Comment by philsturgeon** [comment]
@jerzyn if you copy and paste the commands from my comment above it will work.

**Comment by philsturgeon** [comment]
We have a solution for the first group of errors at the top, we just need to have it ignore `x-summary` like you said. @P0lip is on the case there, and I am looking into why the valid-example error is failing. 

Honestly I am not sure what the issue was in v1.x branch which was causing `hould NOT have additional properties: 0` but I am not sure that it matters as v2.0 is working (just a little noisy right now).

**Comment by P0lip** [comment]
@jerzyn 
I found the cause of the issue on `x-summary` field.
Posted the explanation here https://github.com/stoplightio/spectral/pull/111#issuecomment-477138102

**Comment by jerzyn** [comment]
cool, looks great. thanks @P0lip. My 2c here - agree with what you said in #111 . Spectral should verify whether OAS2 specification elements and hierarchies are valid and ignore anything extra.

**Comment by philsturgeon** [comment]
Merged that fix into the release/2.0 branch! 🎺 🎊🎈

Been in front of a screen all day, gonna pop out for 2 hours then get back to fixing the valid-example errors.

**Comment by philsturgeon** [comment]
Alrighty, got the 2nd valid-example replicated in my tests. 

```
  console.warn src/runner.ts:76
    Encountered error when running rule 'valid-example' on node at path '$,definitions,halRoot':
    Error: can't resolve reference #/definitions/halResource from id #
```

**Comment by philsturgeon** [comment]
@P0lip and I managed to get the valid-example errors fixed: #115.

Thank you for giving us these examples, we managed to improve Spectral a lot with these two example errors. :) 

The release/2.0 branch will be merged later today, and you'll know its out because this PR should automatically close. Or you can subscribe to #112. 

I did notice that spectral has found some issues with your examples, it is suggesting they are not valid. The default CLI output does not give a huuuuge amount of information on this (JSON Schema errors are hard) but if you pull the latest from release/2.0 and run the following, you should get enough information to ensure your examples are valid.

``` bash
yarn setup && yarn build 
./dist/bin/run lint wishlist.yaml --format=json
```

that gives me this output:

```
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halRoot",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halProducts",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halProduct",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"productPatch"
		],
		"message": "should have required property 'title'",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halReviews",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halReview",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halReviewsStats",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halReviewStatsItem"
		],
		"message": "should have required property 'comment'",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halUser",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"user"
		],
		"message": "should have required property 'password'",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halWishlist",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	},
	{
		"code": "valid-example",
		"path": [
			"definitions",
			"halWishProduct",
			"_links",
			"self"
		],
		"message": "should be array",
		"severity": 1
	}
]
```

I think something somewhere is meant to be an array but it is not. :) 

I'll leave you to sort out those bits in your spec. If you think they are incorrect, let me know.

**Comment by jerzyn** [comment]
Ok, cool. Thanks. This spec is just a demo, but maybe worth looking into it :) Interestingly none other validation tool ever complained about this. Will take a look. Thanks for solving this so quickly 👍 I'm also happy if that helped improving spectral!

**Comment by philsturgeon** [comment]
Thank you! Most validators do not bother to check to see if the example is valid, which is why we added the rule :D
